### ISSUE TITLE
Replicate, Copy & Paste

### ISSUE THREAD

**Issue by markadrake** [umco/umbraco-stacked-content#25]
I have 2 features I'd like to request, which could boil down to just one if necessary.

**Replicate** - create a duplication of the current stacked content item.

**Copy & Paste** - copy the model for a stacked content item into localStorage (cookie as fallback) and allow users to paste in the same, or another content node.

These are features that, if not having just moved away from Grid Layout, I would be advocating to make in Umbraco Core. The functionality isn't complex to implement either. I was able to write a proof of concept in just a few minutes with only 2 caveats: 

1. The user interface I'm showing below could be greatly improved.
2. Although `localStorage` is essentially everywhere, we should fallback to a cookie. 

_I looked at going the clipboard route, but it's complex since the click event handler triggers the edit dialog and programmatic access to the clipboard is iffy between user agents._

[![Umbraco - Stacked Content - New Feature](https://img.youtube.com/vi/IB9JcJ7lLEc/0.jpg)](https://www.youtube.com/watch?v=IB9JcJ7lLEc)

**Comment by markadrake** [comment]
## stackedcontent.html

Just add a few icons to the toolbar here. This is the part that I'd want to cleanup.

```
	<a ng-click="prompts[itm.key] = true" class="stack__button" title="Delete"><i class="icon icon-trash"></i></a>
	<a ng-click="copyContent($event, $index)" class="stack__button" title="Duplicate"><i class="icon icon-repeat"></i></a>
	<a ng-click="copyToLocalStorage($event, $index)" class="stack__button" title="Copy"><i class="icon icon-usk-cloud"></i></a>
	<a ng-click="copyFromLocalStorage($event, $index)" class="stack__button" title="Paste"><i class="icon icon-usk-cloud"></i></a>
```

## stackedcontent.controllers.js

Here we just add a few supporting functions for our buttons. This would require some more work before officially accepting - such as the cookie fallback and the paste button shouldn't be active if we reached our max count.

```
	$scope.copyContent = function (evt, idx) {
		var stackedContentItem = Object.assign({}, $scope.model.value[idx]);
		stackedContentItem.key = "";
		stackedContentItem.$$hashKey = "";

		$scope.overlayConfig.event = evt;
		$scope.overlayConfig.data = { model: stackedContentItem, idx: idx + 1, action: "add" };
		$scope.overlayConfig.show = true;
	}

	$scope.copyToLocalStorage = function (evt, idx) {
		var stackedContentItem = Object.assign({}, $scope.model.value[idx]);
		stackedContentItem.key = "";
		stackedContentItem.$$hashKey = "";
		console.log(stackedContentItem);
		window.localStorage.setItem("StackedContentCopy", JSON.stringify(stackedContentItem));
		validateModel(stackedContentItem);
	}

	$scope.copyFromLocalStorage = function (evt, idx) {
		var stackedContentItem = JSON.parse(window.localStorage.getItem("StackedContentCopy"));
		if (validateModel(stackedContentItem)) {
			$scope.overlayConfig.event = evt;
			$scope.overlayConfig.data = { model: stackedContentItem, idx: idx + 1, action: "add" };
			$scope.overlayConfig.show = true;
		}
	}

	// ...
	// Helpers
	// ...

	var validateModel = function (model) {
		try {
			console.log($scope.model.config.contentTypes);
			if (!model || !model.icContentTypeAlias) return false;
			if (!$scope.model.config.contentTypes.filter(x => x.icContentTypeAlias === model.icContentTypeAlias).length) return false;
			console.log("valid model!");
			return true;
		} catch (e) {
			return false;
		}
	}
```

**Comment by mattbrailsford** [comment]
TLDR; Nice feature, needs a little work

My initial reaction is I really don't like the "replicate" workflow as it's really confusing that when you click the replicate button, it opens an edit dialog and then save below which in your example, isn't obvious because the preview is too long so you don't see the new item being created.

My thinking is maybe that it all can be done with copy + paste, so maybe we have a copy icon in the "toolbox" that just copies to the clipboard / local storage and doesn't open a dialog, then, next to the plus icons that appear between items we show a new paste icon (only shows if local storage is populated) and clicking that button pastes in the position. 

I think this workflow is much more inline with what people expect copy / paste to be.

Thoughts?

**Comment by leekelleher** [comment]
I'd prefer a copy-paste feature (as opposed to replicate)! 💯 

@markadrake - code-wise, the code looks clean, (not sure why, but I was expecting a lot more code for this). 👍 

@mattbrailsford - Good suggestion with the "paste" button being next to the "add" (plus) button, makes sense to me.

**Comment by markadrake** [comment]
I agree wholeheartedly with you @mattbrailsford - I think there are now too many icons and the **replicate** is made unnecessary by copy and paste. I actually worked on replicate, was surprised how easy it was and immediately jumped into the copy and paste code.

Thank you @leekelleher  - but its all because of the hard work you guys have put into your projects that I was able to make this so quickly. Happy I'm able to contribute in a meaningful way.

Would you like me to put together a PR, or a few screens to iron out the concept first?

**Comment by mattbrailsford** [comment]
Hey Mark,

Happy for you to put a PR together. Just one thing, I can see in your
screen cast you have your icons permanently showing which I know we
discussed elsewhere, but with respect to the PR please ensure it's only one
change being submitted.

In terms of implementation, I have had a few thoughts 1) on copy, maybe
show a system notification so people know it's copied 2) on paste you'll
need to check allowed do types in case people are copying across instances,
so should only paste if doc type is allowed 3) I can't recall when the key
prop is set, so probably need to double check clearing that in the copy
doesn't cause a problem.

Also, it's whether any of the copy functionality is generic enough that it
can be put into inner content rather than stacked content so it can be used
by other prop editors based upon it?

Other than that, feel free to put a PR together

Many thanks

Matt

On 16 Jan 2018 11:34 pm, "Mark Drake" <notifications@github.com> wrote:

I agree wholeheartedly with you @mattbrailsford
<https://github.com/mattbrailsford> - I think there are now too many icons
and the *replicate* is made unnecessary by copy and paste. I actually
worked on replicate, was surprised how easy it was and immediately jumped
into the copy and paste code.

Thank you @leekelleher <https://github.com/leekelleher> - but its all
because of the hard work you guys have put into your projects that I was
able to make this so quickly. Happy I'm able to contribute in a meaningful
way.

Would you like me to put together a PR, or a few screens to iron out the
concept first?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/umco/umbraco-stacked-content/issues/25#issuecomment-358143798>,
or mute the thread

**Comment by markadrake** [comment]
Okay - be kind - first PR. Please let me know if I can make any additional changes and support this feature in any way. Thanks!

**Comment by leekelleher** [comment]
Thanks @markadrake! 👍 

Closing this ticket off, discussion about the implementation can be had on PR #27
